### DISCUSSION TITLE
Logging messages from language server in language client

### DISCUSSION THREAD

**Discussion by lars-reimann** [microsoft/vscode-discussions#1149] [Extension Development QnA] [answered]
I want to log some messages from a language server in a language client, VS Code specifically. I'm using the methods of [`RemoteConsole`](https://github.com/microsoft/vscode-languageserver-node/blob/8e625564b531da607859b8cb982abb7cdb2fbe2e/server/src/common/server.ts#L133) for this.

In the language client, I configure an output channel:

```ts
const clientOptions: LanguageClientOptions = {
    // ...
    outputChannel: vscode.window.createOutputChannel('Safe-DS', { log: true }),
};

return new LanguageClient('safe-ds', 'Safe-DS', serverOptions, clientOptions);
```

This works well in general, an output channel is created and all log messages show up. The presentation is not good, however:

![image](https://github.com/microsoft/vscode-discussions/assets/2501322/a89792cb-2179-422e-8108-4d90137d1d98)

1. There are two timestamps.
2. There are two message types.
3. Everything that is sent over, is classified as info, so the message types don't even agree.
4. Because of this, message filtering does not work.

I've now changed the configuration of `outputChannel` to 

```
outputChannel: vscode.window.createOutputChannel('Safe-DS', 'log'),
```

This fixes problems 1-3, but since it's no longer a `LogOutputChannel`, message filtering is not offered at all. Is there a better way to handle this?

**Comment by lars-reimann** [comment, marked as answer]
I've now created my own implementation of `LogOutputChannel` with the following `appendLine` method:

```ts
const TRACE_PREFIX = /^\[Trace.*?\] /iu;
const DEBUG_PREFIX = /^\[Debug.*?\] /iu;
const INFO_PREFIX = /^\[Info.*?\] /iu;
const WARN_PREFIX = /^\[Warn.*?\] /iu;
const ERROR_PREFIX = /^\[Error.*?\] /iu;

export class CustomLogOutputChannel implements LogOutputChannel {
    private readonly delegate: LogOutputChannel;

    constructor(name: string) {
        this.delegate = vscode.window.createOutputChannel(name, { log: true });
    }

    appendLine(value: string): void {
        if (TRACE_PREFIX.test(value)) {
            this.delegate.trace(value.replace(TRACE_PREFIX, ''));
        } else if (DEBUG_PREFIX.test(value)) {
            this.delegate.debug(value.replace(DEBUG_PREFIX, ''));
        } else if (INFO_PREFIX.test(value)) {
            this.delegate.info(value.replace(INFO_PREFIX, ''));
        } else if (WARN_PREFIX.test(value)) {
            this.delegate.warn(value.replace(WARN_PREFIX, ''));
        } else if (ERROR_PREFIX.test(value)) {
            this.delegate.error(value.replace(ERROR_PREFIX, ''));
        } else {
            this.delegate.appendLine(value);
        }
    }
}
```

All other methods just call the corresponding method on the `delegate`. Now everything works as expected. It's still a little weird that `appendLine` is always called, regardless of the message type, even if a log output channel is provided.

**Reply by JustinGrote** [reply]
The reason is because the languageServer interface is still OutputChannel, not LogOutputChannel. OutputChannel had no concepts of severity. 

That's being tracked here: https://github.com/microsoft/vscode-languageserver-node/issues/1116

I did something similar to intercept the logs and replace the prefixes, it's the only recourse for now short of using a middleware, which has limitations as I noted in the issue above.
